gc/structs/ForwardedHeader.cpp

```diff
@@ -185,4 +229,115 @@ MM_ForwardedHeader::restoreSelfForwardedPointer()
 	objectHeader->slot = newValue;
 }
 
+
+uintptr_t
+MM_ForwardedHeader::winObjectSectionToCopy(volatile fomrobject_t *copyProgressSlot, fomrobject_t oldValue, uintptr_t *remainingSizeToCopy, uintptr_t outstandingCopies)
+{
+	/* take small section (about 1%) to copy now to maximize parallelism */
+	uintptr_t sizeToCopy = SIZE_OF_SECTION_TO_COPY(*remainingSizeToCopy) & ~_copySizeAlignement;
+	/* but obey minimum increment size (to hide overhead associated with each section) */
+	sizeToCopy = OMR_MAX(sizeToCopy, _minIncrement);
+	/* don't copy more then the remaining size */
+	sizeToCopy = OMR_MIN(*remainingSizeToCopy, sizeToCopy);
+	*remainingSizeToCopy -= sizeToCopy;
+
+	/* atomically try to win this section */
+	fomrobject_t newValue = (fomrobject_t)(*remainingSizeToCopy | (outstandingCopies << OUTSTANDING_COPIES_SHIFT)) | _beingCopiedTag;
+	if (oldValue == lockCompareExchangeObjectHeader(copyProgressSlot, oldValue, newValue)) {
+		return sizeToCopy;
+	}
+
+	return 0;
+}
+
+void
+MM_ForwardedHeader::wait(uintptr_t *spinCount) {
+	/* TODO: consider yielding and/or sleeping, for high spinCount value */
+	for (uintptr_t i = *spinCount; i > 0; i--)	{
+		MM_AtomicOperations::nop();
```

Comment:
This doesn't get optimized away as an empty loop?

Reply:
The very same spin mechanism is used by omrgc_spinlock_acquire.

Reply:
Ok thanks.